File: arm_cpu_simulation.go
package main

import "fmt"

// Register dosyası (R0-R15)
var registers [16]uint32

// ALU işlevleri
func ALU(A, B uint32, ALUControl uint8) uint32 {
    switch ALUControl {
    case 0:
        return A + B // Toplama
    case 1:
        return A - B // Çıkarma
    case 2:
        return A & B // VE
    case 3:
        return A | B // VEYA
    case 4:
        return A ^ B // XOR
    case 5:
        return ^(A | B) // NOR
    default:
        return 0 // Varsayılan durum
    }
}

// Instruction Fetch Decode Execute (IFDE) aşamaları
func executeInstruction(instruction uint32) {
    opcode := instruction >> 26    // Opcode (ilk 6 bit)
    operands := instruction & 0x3FF // Operands (sonraki 10 bit)
    switch opcode {
    case 0: // Toplama
        ALU(registers[0], registers[1], 0)
    case 1: // Çıkarma
        ALU(registers[0], registers[1], 1)
    default:
        fmt.Println("Bilinmeyen Komut")
    }
}

func main() {
    // Test: ALU ve işlemci kullanımı
    registers[0] = 5
    registers[1] = 3
    instruction := uint32(0x20000000) // Örnek komut (toplama)
    executeInstruction(instruction)

    fmt.Printf("ALU Sonucu: %d\n", registers[0]) // Sonuç: 8
}

File: arm_cpu_simulation.rs
use std::fmt;

// Register dosyası (R0-R15)
const NUM_REGISTERS: usize = 16;
static mut REGISTERS: [u32; NUM_REGISTERS] = [0; NUM_REGISTERS];

// ALU işlevleri
fn ALU(A: u32, B: u32, ALUControl: u8) -> u32 {
    match ALUControl {
        0 => A + B,  // Toplama
        1 => A - B,  // Çıkarma
        2 => A & B,  // VE
        3 => A | B,  // VEYA
        4 => A ^ B,  // XOR
        5 => !(A | B), // NOR
        _ => 0,      // Varsayılan durum
    }
}

// CPU'nun temel işlevlerini simüle etmek
fn execute_instruction(instruction: u32) {
    let opcode = instruction >> 26;   // Opcode (ilk 6 bit)
    let operands = instruction & 0x3FF; // Operands (sonraki 10 bit)

    match opcode {
        0 => {
            unsafe {
                REGISTERS[0] = ALU(REGISTERS[0], REGISTERS[1], 0); // Toplama
            }
        },
        1 => {
            unsafe {
                REGISTERS[0] = ALU(REGISTERS[0], REGISTERS[1], 1); // Çıkarma
            }
        },
        _ => println!("Bilinmeyen Komut"),
    }
}

fn main() {
    unsafe {
        REGISTERS[0] = 5;
        REGISTERS[1] = 3;
    }

    let instruction = 0x20000000;  // Örnek komut (toplama)
    execute_instruction(instruction);

    unsafe {
        println!("ALU Sonucu: {}", REGISTERS[0]); // Sonuç: 8
    }
}
